### PR TITLE
PS5 texture updates working.

### PR DESCRIPTION
**Pull request by bbi-michaelanderson**
## Description
This PR includes changes required to get texture updates working properly. Example 08-update now behaves the same as DX11 with the absence of texture blits and compute.

image

### CODE REVIEW

File: src/renderer_agc.h [hunk 1]
```diff
@@ -98,7 +99,29 @@ struct RendererContextAGC final : public RendererContextI
...
+		uint32_t mCount{};
+	};
+
+	// We use a shared pointer to track inflight resources so that if a resource update occurs while a
+	// resource is inflight we can queue the resource to be discarded once it's not inflight and
+	// create a clone with the requested update. The queue used to delete the resource uses the old
+	// inflight counter to determine when it is safe to delete the resource and the updated clone uses
+	// a new counter. Using this method we can simply write updates to an existing resource without discarding
+	// when the resource is not inflight. No double/triple buffering or staging buffers are required as all 
+	// resource memory is either directly updated or cloned and updated by the CPU.
```
Comment (bbi-davidgalloway) [inline src/renderer_agc.h:114]: Seems well reasoned. Will have to see how it works IRL. ;-)

File: examples/08-update/update.cpp [hunk 2]
```diff
@@ -652,11 +652,11 @@ class ExampleUpdate : public entry::AppI
 					const uint16_t tx = m_rng.gen()%(kTexture2dSize-tw);
 					const uint16_t ty = m_rng.gen()%(kTexture2dSize-th);
 
-					uint8_t* dst = &m_texture2dData[(ty*kTexture2dSize+tx)*4];
+					uint8_t* dst = &m_texture2dData[(ty*pitch)+(tx*4)];
 					uint8_t* next = dst + pitch;
 
 					// Using makeRef to pass texture memory without copying.
-					const bgfx::Memory* mem = bgfx::makeRef(dst, tw*th*4);
+					const bgfx::Memory* mem = bgfx::makeRef(dst, pitch*th);
```
Comment (bbi-davidgalloway) [inline examples/08-update/update.cpp:660]: Wondering about the * 4 here.

File: src/renderer_agc.h [hunk 3]
```diff
@@ -158,29 +176,32 @@ struct RendererContextAGC final : public RendererContextI
...
+		FrameBuffer();
 		bool create(uint8_t const num, const Attachment* const attachment);
 		void destroy();
 
-		TextureHandle mColorHandles[BGFX_CONFIG_MAX_FRAME_BUFFER_ATTACHMENTS]{};
-		sce::Agc::CxRenderTarget mColorTargets[BGFX_CONFIG_MAX_FRAME_BUFFER_ATTACHMENTS]{};
-		TextureHandle mDepthHandle{ kInvalidHandle };
+		std::array<TextureHandle, BGFX_CONFIG_MAX_FRAME_BUFFER_ATTACHMENTS> mColorHandles{};
+		std::array<sce::Agc::CxRenderTarget, BGFX_CONFIG_MAX_FRAME_BUFFER_ATTACHMENTS> mColorTargets{};
+		TextureHandle mDepthHandle{kInvalidHandle};
```
Comment (bbi-davidgalloway) [inline src/renderer_agc.h:204]: Why did you change these to std::array?

File: src/renderer_agc.h [hunk 4]
```diff
@@ -279,23 +302,24 @@ struct RendererContextAGC final : public RendererContextI
...
-	Texture mTextures[BGFX_CONFIG_MAX_TEXTURES + 1]{};
-	FrameBuffer mFrameBuffers[BGFX_CONFIG_MAX_FRAME_BUFFERS + 1]{};
+	std::array<IndexBuffer, BGFX_CONFIG_MAX_INDEX_BUFFERS + 1> mIndexBuffers{};
+	std::array<VertexBuffer,BGFX_CONFIG_MAX_VERTEX_BUFFERS + 1> mVertexBuffers{};
+	std::array<Shader, BGFX_CONFIG_MAX_VERTEX_BUFFERS + 1> mShaders{};
+	std::array<Program, BGFX_CONFIG_MAX_PROGRAMS + 1> mPrograms{};
+	std::array<VertexLayout, BGFX_CONFIG_MAX_VERTEX_LAYOUTS + 1> mVertexLayouts{};
+	std::array<Texture, BGFX_CONFIG_MAX_TEXTURES + 1> mTextures{};
+	std::array<FrameBuffer, BGFX_CONFIG_MAX_FRAME_BUFFERS + 1> mFrameBuffers{};
 	std::vector<std::function<bool()>> mDestroyList;
```
Comment (bbi-davidgalloway) [inline src/renderer_agc.h:317]: Same comment. Why std::array?

File: src/renderer_agc.cpp [hunk 5]
```diff
@@ -63,52 +63,6 @@ void freeDmem(void* memVA)
...
-		arraySlice);
-	if (ret != SCE_OK) {
-		return false;
-	}
-
-	return SCE_OK;
-}
-
-//=============================================================================================
-
```
Comment (bbi-davidgalloway) [inline src/renderer_agc.cpp:111]: Was this code you thought you'd need but never did?

File: src/renderer_agc.cpp [hunk 6]
```diff
@@ -769,12 +751,18 @@ bool RendererContextAGC::Shader::create(const Memory& mem)
 void RendererContextAGC::Shader::destroy()
 {
 	// TODO: (manderson) destroy shader.
+
```
Comment (bbi-davidgalloway) [inline src/renderer_agc.cpp:754]: remove todo?

File: src/renderer_agc.cpp [hunk 7]
```diff
@@ -785,46 +773,69 @@ bool RendererContextAGC::Program::create(ShaderHandle const vertexShaderHandle,
 void RendererContextAGC::Program::destroy()
 {
 	// TODO: (manderson) destroy program.
+
```
Comment (bbi-davidgalloway) [inline src/renderer_agc.cpp:776]: Remove todo?

File: examples/08-update/update.cpp (hunk 2 above)
Comment (bbi-michaelanderson) [reply examples/08-update/update.cpp:660]: The * 4 is because its RGBA8 unorm. The problem I fixed is the size of the buffer passed to the update function was smaller than the actual size when taking into consideration the pitch, so the render context had to either ignore it and calculate it's own (which isn't safe, but some of the other render contexts do it) or respect it and fail, in the AGC renderer I respect it (it is supplied by the caller afterall) and fail if it's too small.

File: src/renderer_agc.cpp [hunk 8]
```diff
@@ -785,46 +773,69 @@ bool RendererContextAGC::Program::create(ShaderHandle const vertexShaderHandle,
...
 
 //=============================================================================================
 
 bool RendererContextAGC::Buffer::update(uint32_t const offset, uint32_t const size, const uint8_t* const data)
 {
-	// TODO: (manderson) Once things are up and running consider N buffering to avoid malloc and memcpy, or possibly
-	// staging buffer with GPU copy?
+	// If the buffer is not inflight and it's size hasn't changed we can update it's contents directly,
+	// otherwise we need to modify a copy of the current buffer and discard the current buffer when it's
+	// not inflight.
```
Comment (bbi-davidgalloway) [inline src/renderer_agc.cpp:787]: Can you explain how something is inflight and how it gets in and out of that state?

File: src/renderer_agc.cpp [hunk 9]
```diff
@@ -940,7 +1106,7 @@ bool RendererContextAGC::Texture::create(uint64_t const flags, uint32_t const si
 		sce::Agc::Core::TypedFormat const typedFormat = srgb ? tf.m_fmtSrgb : tf.m_fmt;
 		if (typedFormat == sce::Agc::Core::TypedFormat::kLast)
 		{
-			// TODO: (manderson) Use error texture (pink texture) for missing formats or convert.
+			mTexture.init(); // blank texture.
```
Comment (bbi-davidgalloway) [inline src/renderer_agc.cpp:1109]: IS blank pink?

File: src/renderer_agc.h (hunk 3 above)
Comment (bbi-michaelanderson) [reply src/renderer_agc.h:204]: To get both the compile and runtime bounds checking.

File: src/renderer_agc.cpp (hunk 6 above)
Comment (bbi-michaelanderson) [reply src/renderer_agc.cpp:754]: No I still have to do that.

File: src/renderer_agc.cpp (hunk 5 above)
Comment (bbi-michaelanderson) [reply src/renderer_agc.cpp:111]: No I moved it because I added the texture format conversion to it as well, so it made more sense for it to be a static member of the texture class.

File: src/renderer_agc.cpp (hunk 9 above)
Comment (bbi-michaelanderson) [reply src/renderer_agc.cpp:1109]: No, it's black at the moment. I have plans to swap that with a default pink texture.

File: src/renderer_agc.cpp (hunk 8 above)
Comment (bbi-michaelanderson) [reply src/renderer_agc.cpp:787]: Inflight means the resource was used in a frame that has been dispatched to the GPU, once the frame is completed (we are about to reuse it's context) and we know the GPU is done with it I decrement the inflight counter, if it's 0 I know it's not in use by any pending frame so I can safely manipulate the resources memory directly.

Comment (bbi-michaelanderson) [reply src/renderer_agc.cpp:787]: When I attempt to update a resource and it's inflight (counter is > 0) then I create a copy with the update and add the existing buffer to a queue that will free the buffer once the inflight counter reaches zero. Passing the counters around as shared pointers allows me to simply create a new counter (with count = 0) if I copy the buffer, that way if another update is issued before the new buffer is used I can simply update the new buffer directly. All of this basically helps encourage direct manipulation of the data when possible.

Comment (bbi-michaelanderson) [reply src/renderer_agc.cpp:787]: Vertex and index buffers are a bit smarter since we only need a buffer size that satisfies the largest update (and they are updated in linear blocks), so as more of the buffer is updated I reallocate it, this helps avoid allocating the huge transient vertex and index buffers during bgfx init, I only allocate what the application actually uses. I really like the AGC approach to resource memory, all of the resources are basic buffers that when used you simply supply a struct that describes it's layout/usage.
